### ISSUE TITLE
rule ideas

### ISSUE THREAD

**Issue by waffledonkey** [yaniswang/HTMLHint#71]
**reasonably important**
- I wouldn't mind checks for page missing ```<title>``` 
-  ```<script>``` before ```<link rel=stylesheet>``` has known performance problems. ```head-script-disabled``` works well, but there are some scripts that absolutely must be in the head (html5shiv, various polyfills, etc.) so running ```head-script-disabled``` is not always practical.

**personal opinion**
being unkind to inline *```style```* attributes as well as ```<style>``` blocks, but I'm not sure it's appropriate to roll them up under the same ```style-disabled``` rule.

**Comment by yaniswang** [comment]
What's your suggest?

**Comment by waffledonkey** [comment]
I'm not sure they'll meet with your naming or coding style conventions and I didn't focus too much on making sure line or column numbers were right, but I've prepared the following -- which regrettably I cannot attach.

Regards,

**attr-no-style.js**
```
HTMLHint.addRule({
    id: 'attr-no-style',
    description: 'style attributes are not allowed.',
    init: function(parser, reporter){
        var self = this;
        parser.addListener('tagstart', function(event){
            var tagName = event.tagName.toLowerCase(),
                mapAttrs = parser.getMapAttrs(event.attrs),
                col = event.col + tagName.length + 1;
            if('style' in mapAttrs){
                reporter.warn(self.description, event.line, col, self, event.raw);
            }
        });
    }
});
```

**head-script-some.js**
```
HTMLHint.addRule({
    id: 'head-script-some',
    description: 'script tags should come after link[rel=stylesheet] in head',
    init: function(parser, reporter){
        var self = this,
            found = false,
            reScript = /^(text\/javascript|application\/javascript)$/i;

        var allEvent = function(event){
            var mapAttrs, type;

            if(event.type === 'tagstart' && event.tagName.toLowerCase() === 'script'){
                mapAttrs = parser.getMapAttrs(event.attrs),
                type = mapAttrs.type;
                if (!type || reScript.test(type) === true){
                    found = event.line;
                }
            }
            if(event.type === 'text' && found !== false && event.line > found && /link/.test(event.raw)){
                reporter.warn(self.description, event.line + 1, null, self, event.raw);
            }
        };
        function onTagEnd(event){
            if(event.tagName.toLowerCase() === 'head'){
                parser.removeListener('all', allEvent);
                parser.removeListener('tagend', onTagEnd);
            }
        }
        parser.addListener('all', allEvent);
        parser.addListener('tagend', onTagEnd);
    }
});
```

**title-require.js**
```
HTMLHint.addRule({
    id: 'title-require',
    description: 'the title tag must be present in head.',
    init: function(parser, reporter){
        var self = this,
            ok = false;

        function onTagStart(event){
            if(event.tagName.toLowerCase() === 'title'){
                ok = true;
                parser.removeListener('tagstart');
            }
        }
        function onTagEnd(event){
            if(event.tagName.toLowerCase() === 'head'){
                if (!ok) {
                    reporter.warn(self.description, event.line, event.col, self, event.raw);
                }
                parser.removeListener('tagstart', onTagStart);
                parser.removeListener('tagstart', onTagEnd);
            }
        }
        parser.addListener('tagstart', onTagStart);
        parser.addListener('tagend', onTagEnd);
    }
});
```

**Comment by yaniswang** [comment]
rule title-require added: https://github.com/yaniswang/HTMLHint/commit/2247853885904388ad8e550b5ae105a04a5d7731

The other 2 rules wait more messages.

**Comment by waffledonkey** [comment]
Thank you very much

**Comment by yaniswang** [comment]
add rule: `inline-style-disabled`

https://github.com/yaniswang/HTMLHint/blob/master/src/rules/inline-style-disabled.js

**Comment by monish-it** [comment]
> .getMapAttrs(event.attrs)

what does function .getMapattrs() do? 
Pls share any resources if available.

**Comment by waffledonkey** [comment]
Damn that was a long time ago.  Based on a cursory examination of what I had written then, you call addRule (a method against the HTMLLint object) with an object consisting of an id, a description and an init callback which will have at it's disposal the parser and reporter object instances respectively.  mapAttrs is just a variable name I chose at the time, it holds the attributes of the tag being parsed.  So if html lint was parsing <h1 class="red"> mapAttrs would hold {class: "red"}.

What I was attempting to do is identify if any <script> tags that occurred before <link rel="stylesheet"> tags but I used a weak heuristic which doesn't actually test for rel="styleheet" and there are other <link> elements (e.g. <link rel="icon">.  

As to why someone would want to know that or what the big deal is see [this answer](https://stackoverflow.com/a/9272629) from Steve Souders who was -- at that time, I don't know what he's doing these days -- lead performance engineer at Google, Velocity conference co-founder and author of a couple of books on the subject of web performance. I'm sure he has other accolades, the dude was prolific.

Modern browsers and reactive frameworks -- if that's your thing -- may change things... 

Good luck
